File: backend/apps/accounts/permissions.py
"""
Permissões por perfil e por NÍVEL — item 2.1.1.6 e hierarquia definida na
reunião de 22/07/2026 (Master / Sênior / Pleno / Júnior, nos ambientes
BackEnd (equipe interna) e FrontEnd (cliente)).

Matriz vigente (padrão inicial — o cliente ainda vai detalhar Sênior/Pleno/Júnior):
    Master  → cura os dados de sistema, gerencia usuários e exclui.
    Sênior  → opera (coletas/laudos/OSPs) e exclui; não mexe em dados de sistema.
    Pleno   → opera; NÃO exclui; não mexe em dados de sistema.
    Júnior  → opera; NÃO exclui; não mexe em dados de sistema.
"""
from rest_framework.permissions import BasePermission, SAFE_METHODS

from .models import PERFIS_INTERNOS, Perfil


class IsInterno(BasePermission):
    """Apenas equipe da CONTRATADA (Admin/Gestor/Técnico)."""

    message = "Apenas usuários internos têm acesso a este recurso."

    def has_permission(self, request, view):
        return bool(request.user and request.user.is_authenticated and request.user.is_interno)


class IsAdmin(BasePermission):
    message = "Apenas administradores."

    def has_permission(self, request, view):
        return bool(
            request.user and request.user.is_authenticated and request.user.perfil == Perfil.ADMIN
        )


class IsMaster(BasePermission):
    """Apenas Master interno; Master de cliente nunca administra a plataforma."""

    message = "Apenas usuários de nível Master podem executar esta ação."

    def has_permission(self, request, view):
        user = request.user
        return bool(user and user.is_authenticated and user.is_interno and user.is_master)


class InternoEditaClienteVisualiza(BasePermission):
    """
    Interno: leitura+escrita. Cliente: somente leitura (Portal — item 2.7).
    Exclusão exige nível Master ou Sênior — Júnior/Pleno alimentam mas não apagam.

    É o padrão do Portal: vale para o parque (Cliente/Área/Setor/Equipamento/
    Componente/Dados técnicos/Rota), para os recursos técnicos (inspeções,
    medições, análises, laudos) e para os catálogos internos (Prestador,
    Instrumentação, dados de sistema). O único recurso em que o cliente escreve
    é a Ordem de Serviço, e só nos campos de retorno — ver
    `InternoEditaClienteResponde`.
    """

    def has_permission(self, request, view):
        user = request.user
        if not (user and user.is_authenticated):
            return False
        if request.method in SAFE_METHODS:
            return True
        if user.perfil not in PERFIS_INTERNOS:
            return False
        if request.method == "DELETE" and not user.pode_excluir:
            self.message = "Seu nível de acesso não permite excluir registros."
            return False
        return True


class InternoEditaClienteResponde(BasePermission):
    """
    Interno: leitura+escrita, exclusão exige nível Master ou Sênior (igual
    `InternoEditaClienteVisualiza`).

    Cliente: consulta + RETORNO DE INFORMAÇÃO. Ele não cria nem exclui nada —
    só atualiza (PATCH/PUT) um registro que a CONTRATADA já abriu, e apenas nos
    campos de resposta. QUAIS campos é decidido no serializer
    (`OrdemServicoSerializer.CAMPOS_RETORNO_CLIENTE`), não aqui: esta classe só
    responde "pode tocar neste tipo de recurso, e com qual verbo".

    Decisão de 2026-09-21, que substitui a de 2026-09-18: o Portal volta a ser
    somente leitura em todo o resto (parque, rotas, inspeções, medições,
    análises), como no item 2.7 do Anexo I. A única escrita que sobra é a que o
    cliente é a ÚNICA fonte: quando a corretiva foi planejada e executada, quem
    executou, o que foi feito, se a abertura da máquina confirmou o
    diagnóstico, e quanto custou de fato — números que alimentam o ROI da
    Seção D e o KPI de acerto do diagnóstico.

    O ESCOPO por cliente continua em `get_queryset()` de cada ViewSet
    (`escopo_cliente`/`campo_cliente`) — esta classe nunca decide "em qual
    cliente".
    """

    #: Verbos de retorno: alteram um registro existente, nunca criam nem apagam.
    METODOS_RETORNO = {"PATCH", "PUT"}

    def has_permission(self, request, view):
        user = request.user
        if not (user and user.is_authenticated):
            return False
        if request.method in SAFE_METHODS:
            return True
        if user.is_interno:
            if request.method == "DELETE" and not user.pode_excluir:
                self.message = "Seu nível de acesso não permite excluir registros."
                return False
            return True
        if user.is_cliente and request.method in self.METODOS_RETORNO:
            return True
        self.message = (
            "No Portal você consulta e devolve as informações da execução. "
            "Criar ou excluir registros é feito pela equipe da ThermoProActive."
        )
        return False


class MasterEditaDemaisVisualizam(BasePermission):
    """
    Só o Master (interno) cria/edita/remove; os demais autenticados apenas leem.
    Usada nas tabelas de referência (dados de sistema): o cliente definiu que
    Sênior/Pleno/Júnior precisam solicitar ao Master para inserir esses cadastros,
    o que evita divergência de padronização e exclusões não autorizadas.
    """

    message = "Apenas o nível Master pode alterar os dados de sistema."

    def has_permission(self, request, view):
        user = request.user
        if not (user and user.is_authenticated):
            return False
        if request.method in SAFE_METHODS:
            return True
        return user.pode_curar_dados_sistema


class SouDoCliente(BasePermission):
    """
    Portal: qualquer usuário do cliente vê a própria equipe; só o Master do
    lado do cliente convida (mesma regra de PodeConvidar, em ConviteView).
    """

    message = "Apenas usuários do Portal do Cliente têm acesso a este recurso."

    def has_permission(self, request, view):
        user = request.user
        if not (user and user.is_authenticated and user.is_cliente and user.cliente_id):
            return False
        if request.method in SAFE_METHODS:
            return True
        self.message = "Apenas o Master da sua organização pode convidar colaboradores."
        return user.is_master

File: frontend/src/lib/permissions.ts
"use client";

import { useMemo } from "react";
import { useAuth } from "./auth";
import type { Nivel, Perfil, User } from "./types";

/**
 * RBAC do front-end — espelho fiel das permissões do backend.
 *
 * IMPORTANTE (segurança): isto é UX, não é controle de acesso. A autorização
 * real vive no DRF (`apps/accounts/permissions.py`: IsInterno, IsMaster,
 * InternoEditaClienteVisualiza, MasterEditaDemaisVisualizam). O front usa este
 * módulo para NÃO OFERECER o que o usuário não pode fazer — evitando o 403 na
 * cara do usuário — e para desviar rotas que ele não deveria alcançar.
 *
 * Se uma regra mudar no backend, mude aqui também: os dois lados são a mesma
 * matriz, decidida com o cliente (Anexo I 2.1 + reunião de 22/07/2026).
 *
 *   Ambiente × Nível
 *   BackEnd (equipe interna)      FrontEnd (cliente)
 *     Master  cura dados de sistema, gerencia usuários, exclui
 *     Sênior  opera e exclui
 *     Pleno   opera, não exclui
 *     Júnior  opera, não exclui
 *
 * O FrontEnd (Portal) é de CONSULTA. A única escrita que ele tem é o retorno
 * de informação da Ordem de Serviço (`osp:responder`) — decisão de
 * 2026-09-21, que reverte a de 2026-09-18. Convidar a própria equipe
 * (`equipe:gerenciar`) continua com o Master do cliente: é gestão da conta,
 * não dado operacional.
 */

/** Cada verbo que a interface precisa decidir mostrar ou não. */
export type Habilidade =
  /* Áreas (qual portal o usuário enxerga) */
  | "area:admin"
  | "area:portal"
  /* Operação sobre dados do cliente */
  | "operacional:ver"
  | "operacional:criar"
  | "operacional:editar"
  | "operacional:excluir"
  /* Cadastro de clientes e sua estrutura (áreas, setores, equipamentos, rotas) —
     exclusivo da equipe interna, em /clientes, /equipamentos, /rotas (admin). */
  | "clientes:gerenciar"
  /*
   * Escrita no parque técnico (equipamentos, rotas, inspeções, medições).
   * Exclusivo da equipe interna desde 2026-09-21, quando a decisão de
   * autoatendimento de 2026-09-18 foi revertida: o Portal consulta e devolve
   * informação, mas quem cadastra e diagnostica é a CONTRATADA.
   */
  | "parque:gerenciar"
  /* Tabelas de referência do sistema (normas, tipos, criticidades…) */
  | "dados-sistema:ver"
  | "dados-sistema:curar"
  /* Pessoas */
  | "usuarios:ver"
  | "usuarios:gerenciar"
  /* Equipe do PRÓPRIO cliente, vista de dentro do Portal (não confundir com
     "usuarios:*", que é a concessão de acesso pela ThermoProActive). */
  | "equipe:ver"
  | "equipe:gerenciar"
  /* Documentos técnicos */
  | "laudos:ver"
  | "laudos:emitir"
  | "laudos:assinar"
  /* Ordens de serviço */
  | "osp:ver"
  | "osp:criar"
  | "osp:mudar-status"
  /* Retorno de informação: o cliente devolve quando a corretiva foi planejada
     e executada, por quem, o que foi feito, se o diagnóstico se confirmou e
     quanto custou. É a ÚNICA escrita do Portal — ver, no backend,
     `OrdemServicoSerializer.CAMPOS_RETORNO_CLIENTE`. */
  | "osp:responder"
  /* Relatórios e exportações */
  | "relatorios:ver"
  | "relatorios:exportar"
  /* Prestadores de serviço terceirizados */
  | "prestadores:gerenciar";

const NIVEIS_QUE_EXCLUEM: Nivel[] = ["MASTER", "SENIOR"];

/** Resolve uma habilidade para um usuário. Única fonte da verdade no front. */
export function can(user: User | null | undefined, habilidade: Habilidade): boolean {
  if (!user) return false;

  const interno = user.is_interno;
  const cliente = user.is_cliente;
  // Confia no flag do backend quando ele vem; o nível é só fallback defensivo.
  const excluir = user.pode_excluir ?? NIVEIS_QUE_EXCLUEM.includes(user.nivel);
  const master = user.is_master;

  switch (habilidade) {
    case "area:admin":
      return interno;
    case "area:portal":
      return cliente;

    case "operacional:ver":
      return interno || cliente;
    case "operacional:criar":
    case "operacional:editar":
      return interno;
    case "operacional:excluir":
      return interno && excluir;

    case "clientes:gerenciar":
      return interno;

    case "parque:gerenciar":
      return interno;

    case "dados-sistema:ver":
      return interno;
    case "dados-sistema:curar":
      return user.pode_curar_dados_sistema ?? (interno && master);

    /* Só o Master concede acessos (backend: UserViewSet + IsMaster). Como o
       ViewSet inteiro exige Master, nem a leitura é liberada aos demais. */
    case "usuarios:ver":
    case "usuarios:gerenciar":
      return interno && master;

    /* Ver os colegas da própria empresa: qualquer usuário do Portal. Convidar
       (e um dia gerenciar) exige ser o Master do lado do cliente — é ele quem
       o backend deixa criar convites PORTAL para o próprio `cliente`
       (ConviteView + PodeConvidar). */
    case "equipe:ver":
      return cliente;
    case "equipe:gerenciar":
      return cliente && master;

    case "laudos:ver":
      return interno || cliente;
    case "laudos:emitir":
      return interno;
    /* Assinar exige registro em conselho de classe (CREA) — regra do Anexo I 3.1.2. */
    case "laudos:assinar":
      return interno && !!user.conselho_classe;

    case "osp:ver":
      return interno || cliente;
    case "osp:criar":
    case "osp:mudar-status":
      return interno;
    /* Qualquer nível do Portal responde: quem executou a corretiva costuma ser
       o manutentor, não o gestor. O backend também não exige Master aqui. */
    case "osp:responder":
      return interno || cliente;

    case "relatorios:ver":
    case "relatorios:exportar":
      return interno || cliente;

    case "prestadores:gerenciar":
      return interno;

    default:
      return false;
  }
}

/** Todas as habilidades precisam valer. */
export function canAll(user: User | null | undefined, habilidades: Habilidade[]): boolean {
  return habilidades.every((h) => can(user, h));
}

/** Basta uma habilidade valer. */
export function canAny(user: User | null | undefined, habilidades: Habilidade[]): boolean {
  return habilidades.some((h) => can(user, h));
}

/* ==========================================================================
   Rótulos de perfil — linguagem do usuário, não do banco.
   ========================================================================== */

export const PERFIL_LABEL: Record<Perfil, string> = {
  ADMIN: "Administrador",
  GESTOR: "Gestor / Supervisor",
  TECNICO: "Técnico / Analista",
  CLIENTE_CORP: "Cliente — Gestor Corporativo",
  CLIENTE_LOCAL: "Cliente — Gestor Local",
  CLIENTE_PCM: "Cliente — PCM",
  CLIENTE_MANUT: "Cliente — Manutentor",
};

export const NIVEL_LABEL: Record<Nivel, string> = {
  MASTER: "Master",
  SENIOR: "Sênior",
  PLENO: "Pleno",
  JUNIOR: "Júnior",
};

/** O que cada nível pode, em uma linha — usado na tela de usuários. */
export const NIVEL_RESUMO: Record<Nivel, string> = {
  MASTER: "Cura os dados de sistema, concede acessos e exclui registros.",
  SENIOR: "Opera inspeções, laudos e OSPs, e pode excluir registros.",
  PLENO: "Opera inspeções, laudos e OSPs. Não exclui registros.",
  JUNIOR: "Alimenta o sistema com acompanhamento. Não exclui registros.",
};

export const PERFIS_INTERNOS: Perfil[] = ["ADMIN", "GESTOR", "TECNICO"];
export const PERFIS_CLIENTE: Perfil[] = [
  "CLIENTE_CORP",
  "CLIENTE_LOCAL",
  "CLIENTE_PCM",
  "CLIENTE_MANUT",
];

/* ==========================================================================
   Guarda de rota — mapa declarativo prefixo → habilidade exigida.
   Ordem importa: o primeiro prefixo que casar decide.
   ========================================================================== */

type RegraRota = { prefixo: string; exige: Habilidade };

const REGRAS_ROTA: RegraRota[] = [
  { prefixo: "/portal/equipe", exige: "equipe:ver" },
  // A única tela de escrita do Portal: devolver as informações da execução de
  // uma OSP. Prefixo específico, checado ANTES do "/portal" genérico abaixo.
  // As rotas de cadastro que existiam aqui (/portal/equipamentos/novo,
  // /portal/ordens-servico/novo, /portal/inspecoes/editar) foram removidas
  // junto com as páginas em 2026-09-21.
  { prefixo: "/portal/ordens-servico/editar", exige: "osp:responder" },
  { prefixo: "/portal", exige: "area:portal" },
  { prefixo: "/usuarios", exige: "usuarios:ver" },
  { prefixo: "/cadastros", exige: "dados-sistema:ver" },
  { prefixo: "/clientes", exige: "clientes:gerenciar" },
  { prefixo: "/prestadores", exige: "prestadores:gerenciar" },
  { prefixo: "/equipamentos", exige: "clientes:gerenciar" },
  { prefixo: "/rotas", exige: "clientes:gerenciar" },
  { prefixo: "/inspecoes", exige: "operacional:ver" },
  { prefixo: "/relatorios-inspecao", exige: "operacional:ver" },
  { prefixo: "/servicos", exige: "operacional:ver" },
  { prefixo: "/osps", exige: "osp:ver" },
  { prefixo: "/laudos", exige: "laudos:ver" },
  { prefixo: "/relatorios", exige: "relatorios:ver" },
  { prefixo: "/alertas", exige: "operacional:ver" },
  { prefixo: "/notificacoes", exige: "operacional:ver" },
  { prefixo: "/dashboard", exige: "area:admin" },
];

/** A rota é permitida para este usuário? Rota sem regra é liberada a autenticados. */
export function podeAcessarRota(user: User | null | undefined, pathname: string): boolean {
  if (!user) return false;
  const regra = REGRAS_ROTA.find((r) => pathname === r.prefixo || pathname.startsWith(`${r.prefixo}/`));
  return regra ? can(user, regra.exige) : true;
}

/**
 * Onde este usuário deve cair ao entrar (ou quando bater numa rota proibida).
 * Cliente vai para o portal; equipe interna, para o painel operacional.
 */
export function rotaInicial(user: User | null | undefined): string {
  if (!user) return "/portal/login";
  return user.is_cliente ? "/portal/dashboard" : "/admin/dashboard";
}

/**
 * Rotas antigas de cliente → equivalente no portal. Links salvos pelo cliente
 * (favorito, e-mail antigo) continuam funcionando depois da separação dos portais.
 */
const EQUIVALENTE_NO_PORTAL: Record<string, string> = {
  "/laudos": "/portal/laudos",
  "/inspecoes": "/portal/inspecoes",
  "/osps": "/portal/ordens-servico",
  "/equipamentos": "/portal/equipamentos",
  "/relatorios": "/portal/relatorios",
  "/notificacoes": "/portal/alertas",
  "/dashboard": "/portal",
};

/** Destino de redirecionamento para um cliente que caiu numa rota interna. */
export function destinoNoPortal(pathname: string): string {
  const raiz = `/${pathname.split("/").filter(Boolean)[0] ?? ""}`;
  const base = EQUIVALENTE_NO_PORTAL[raiz];
  if (!base) return "/portal";
  // Preserva o /:id do detalhe (ex.: /laudos/12 → /portal/laudos/12).
  const resto = pathname.slice(raiz.length);
  return resto ? `${base}${resto}` : base;
}

/* ==========================================================================
   Hook de conveniência para componentes.
   ========================================================================== */

export function usePermissoes() {
  const { user } = useAuth();
  return useMemo(
    () => ({
      user,
      pode: (h: Habilidade) => can(user, h),
      podeTodas: (hs: Habilidade[]) => canAll(user, hs),
      podeAlguma: (hs: Habilidade[]) => canAny(user, hs),
      /* Atalhos usados o tempo todo nas telas. */
      interno: !!user?.is_interno,
      ehCliente: !!user?.is_cliente,
      master: !!user?.is_master,
      podeEditar: can(user, "operacional:editar"),
      podeExcluir: can(user, "operacional:excluir"),
    }),
    [user]
  );
}
